chia/data_layer/data_layer_wallet.py

```diff
@@ -736,6 +737,78 @@ async def singleton_removed(self, parent_spend: CoinSpend, height: uint32) -> No
                     self.id(),
                 )
             )
+            await self.check_for_rebase(singleton_record.launcher_id)
+
+    async def check_for_rebase(self, launcher_id: bytes32) -> None:
+        """
+        This method is meant to detect a fork in our expected pending singletons and the singletons that have actually
+        been confirmed on chain.  If there is a fork and the root on chain never changed, we will attempt to rebase our
+        singletons on to the new latest singleton.  If there is a fork and the root changed, we assume that everything
+        has failed and delete any pending state.
+        """
+        unconfirmed_singletons = await self.wallet_state_manager.dl_store.get_unconfirmed_singletons(launcher_id)
+        if len(unconfirmed_singletons) == 0:
+            return
+        unconfirmed_singletons = sorted(unconfirmed_singletons, key=attrgetter("generation"))
+        full_branch: List[SingletonRecord] = await self.wallet_state_manager.dl_store.get_all_singletons_for_launcher(
+            launcher_id,
+            min_generation=unconfirmed_singletons[0].generation,
+        )
+        if len(unconfirmed_singletons) == len(full_branch) and set(unconfirmed_singletons) == set(full_branch):
+            return
+
+        # Now we have detected a fork so we should check whether the root changed at all
+        parent_singleton = await self.wallet_state_manager.dl_store.get_singleton_record(
+            unconfirmed_singletons[0].lineage_proof.parent_name
+        )
+        if parent_singleton is None or {parent_singleton.root} != {s.root for s in full_branch if s.confirmed}:
```

Comment:
I think this may be a mypy thing because we should always have a parent.  This is also helpful during reorgs I think in which the branch isn't even based on a history that exists anymore and therefore has no parent.